### PR TITLE
Chests rework

### LINKED ISSUE #73: enhancing chests
it would be nice to see https://github.com/GreenXenith/technic_chests features added here, sorting, items to and from chest to inv is invaluable for big material moves

### LINKED ISSUE #105: .luacheckrc contains unnecessary ignores
Many of ignored linter checks should fail and not get ignored.

Failing things should be checked in code not by disabling checks.
Some ignores are of course fine but most ignores that target specific source files in current .luacheckrc are not fine at all, currently there's things that can allow a lot of crash causing simple mistakes to slip through.

**Remove these ignores** and only add back ones really needed (almost sure that none of these are really needed) and issue cannot be fixed:
https://github.com/mt-mods/technic/blob/264029a66f06f945e305de106aca6ad8354d5850/.luacheckrc#L32-L47

Also before adding any of those or like those always **consider adding specific ignore rule through source code comments**, that way it is at least visible when code is modified.
https://luacheck.readthedocs.io/en/stable/inline.html

### PR DESCRIPTION
**Pull request by OgelGames**
Improves technic chests with extra features, along with a rewrite of the code.

### Added
- Inventory quick-move functionality. (what was originally requested in #73)
- More ways to sort items.
- `digilines` support. (inventory querying and put/take events)
- Protected versions of all chests.

### Changed
- Redesigned the formspecs of chests.
- All chests have description/infotext editing.
- Protected chests don't require the `protector` or `area` mods.
- Adjusted color overlay textures.

### Removed
- `tubelib` support. (the rest of technic doesn't support it, and it doesn't work well with the new functionality)
- `moreblocks` optional dependency.
- Formspec background.

Fixes #73
Closes #105

@BuckarooBanzay @S-S-X @SwissalpS @6r1d @jessicara @wsor4035

### CODE REVIEW

**Comment by S-S-X** [comment]
> Protected versions of all chests.

Now that this is adding new nodes for each chest would it be possible to reduce nodes and have protection parameters in metadata as metadata is already heavily used by chests and by owner check.
Crafts and required metadata for public/protected/owned checks can be done easily but not sure if adding texture overlay is that easy: protector logo / indicator for protected and lock overlay for owned?

**Comment by SwissalpS** [comment]
I tested the basics, no digiline. I think the interface is ok. Users will get used to it quickly.
Also tested with (ultra)sonic screwdriver and worked fine.

If it is possible to do what SX describes with overlays, that would be cool.

**Comment by OgelGames** [comment]
Node overlays/texture changes are not possible (yet), see https://github.com/minetest/minetest/issues/9193, https://github.com/minetest/minetest/issues/9019, and https://github.com/minetest/minetest/issues/36.

Also, having separate nodes is much simpler, and is also consistent with wooded chests.

BTW, we currently have somewhere over 21k nodes registered, and every node that gets added to the circular saw adds 48 more...

**Comment by S-S-X** [comment]
> Also, having separate nodes is much simpler, and is also consistent with wooded chests.

Probably slightly simpler to make it work yes. About consistency with wooden chests, I don't think it is good to keep consistency like that just because of legacy stuff. I think not much besides having multiple nodes and right click to open is really consistent between wooden chests and technic chests.

Instead interesting thing to know would be which direction minetest is pushing with future updates, will it be more nodes or would it go towards more dynamic nodes.

https://github.com/minetest/minetest/issues/9193 had very good ideas for implementations but seems like it is not yet time to start really utilizing dynamic nodes. I think only possible approach currently would be using entities as overlays and this probably is not good for performance?

> BTW, we currently have somewhere over 21k nodes registered, and every node that gets added to the circular saw adds 48 more...

Yes things like that is one of the reasons why I tried to also think if there would be usable way to reduce nodes when extra nodes are not really needed, so just technic chests would be way more than circular saw support for another node.
Looking at that 21k nodes registered it indeed seems that many other mods also seem to handle things that way and probably because that has been only way to do these things.

Seems like this is not possible this time, entities for thing like this would probably be way too complicated.

**Comment by SwissalpS** [comment]
> entities as overlays and this probably is not good for performance?

Very bad for performance. Also overlaying chests with entities would mean interaction is done through entities.
Cannot recommend this path.

**Comment by OgelGames** [comment]
@S-S-X what I mean by consistency is this:

screenshot_20201104_182907

**Comment by S-S-X** [comment]
> @S-S-X what I mean by consistency is this:

I did not probably really understand what you meant, chests look similar? They do somewhat, technic is just missing open lid versions.
However I did not propose to change visual look or textures at all but instead was looking for better way to implement those visuals without introducing more nodes for textures and possibly for other minor (in terms of LoC) features.
But as you did dig up those engine feature requests it became clear that it is not yet possible and new features allowing lot more flexible visuals are not ready for use yet as only small mostly irrelevant parts have been actually implemented for engine (dynamic SAO textures).

Also issues linked here would address open chest visuals for wooden chests without requiring another node. Better visual consistency with wooden chests could be done by adding fourth variant with open lid but maybe not trade that for another 20 nodes as that would further increase total nodes just for 5 chests to 80 nodes.
Especially when there's approved tickets for engine features meant to solve exactly these issues.

Btw, there was another interesting and related feature request about so called node bound server active object. Apparently simpler and cheaper SAO, like something between full featured mt entity and particle spawner texture sprite. Linked that issue to discord too.

File: technic_chests/init.lua [hunk 1]
```diff
@@ -1,77 +1,74 @@
...
-	output = "default:chest_locked",
-	type = "shapeless",
-	recipe = {
-		"default:chest",
-		"basic_materials:padlock"
-	}
-})
+end)
 
+-- FIXME: Why does this LBM exist?
```
Comment (BuckarooBanzay) [inline technic_chests/init.lua:71]: I think that was for compatibility reasons a long time ago.
The default chest uses a `on_rightclick` sent formspec: https://github.com/minetest/minetest_game/blob/master/mods/default/chests.lua#L210
I guess it wasn't always like this.

**IMO**: remove it if you want, if it turns out to be an issue it should be added in the `default` mod anyway

**Comment by BuckarooBanzay** [comment]
I played around with it on the test-server, compatibility looks good :+1:
The mithril chest interface is pure awesome, i like it!

About the node-count: how much nodes are additionally registered now? It can't be above 10 or 20 i guess.
Anyways: i don't think this should be an issue in this order of magnitude... (IMO)

**Comment by wsor4035** [comment]
looks good, on the matter of extra nodes, could it be simplified down to having a toggle option of being protected or not?

**Comment by S-S-X** [comment]
> having a toggle option of being protected or not?

My opinion would be sacrifice nodes to keep visual indicator outside, possibly rework that part :shrug: after engine offers required bells and whistles..
If that could be done then I think how meta flag for protection could be toggle switch or crafting recipe and that's more just opinion as complexity for both is about similar.

At least I would very much like to see if chest is protected or locked without actually opening formspec and I have feeling that many other players like feature very much too.

Or was that for toggle option **with** node swapping to keep type specific textures?

**Comment by SwissalpS** [comment]
current UX should not be broken. Players should be able to see difference between unlocked, locked and protected chests without formspec or infotext telling them.

File: technic_chests/init.lua (hunk 1 above)
Comment (OgelGames) [reply technic_chests/init.lua:71]: I think I'll remove it, I agree that it doesn't belong here in the first place.

**Comment by OgelGames** [comment]
> About the node-count: how much nodes are additionally registered now? It can't be above 10 or 20 i guess.

Only 19 more, 16 for gold chest, and 1 each for iron, copper, and silver.

**Comment by S-S-X** [comment]
> > About the node-count: how much nodes are additionally registered now? It can't be above 10 or 20 i guess.
> 
> Only 19 more, 16 for gold chest, and 1 each for iron, copper, and silver.

Isn't it exactly 20 or is there something special with mithril chest?

**Review by SwissalpS — COMMENTED** [review]
idk, first guess is that some chests stayed open, but it's only resetting the formspec so maybe some mistake happened and this was to fix it?

**Comment by OgelGames** [comment]
For some reason I didn't think of just looking at commit history...

This is why the LBM existed: https://github.com/minetest-mods/technic/issues/428

**Comment by S-S-X** [comment]
Probably safe to remove that LBM already or move it into some legacy compatibility collection file with separate config option to enable. If removed then it could be good to add note about this to some documentation file / mod install instructions.
LBM that cleans up metadata from old network stuff is another LBM that is not needed if mod is installed on new server.

**Comment by OgelGames** [comment]
I think in this case it's far too old to be necessary, only maps/servers that had technic installed before 2017/06/18 ([this commit](https://github.com/minetest-mods/technic/commit/d4162be9a75816e2edbc8f6ed97588b2cad881d7)) need it at all (not pandorabox), and it's only a visual bug anyway.

**Comment by OgelGames** [comment]
I've done yet more testing (mostly of the digiline commands), and I think this is ready for merging. If anyone wants to do more testing, now is the time to do it :wink:

I'll merge tomorrow if there are no objections and no bugs are found.

**Pull request merged**